runtime/vm/resolvesupport.cpp

```diff
@@ -1869,6 +1869,115 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
 
 }
 
+j9object_t
+resolveConstantDynamic(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpIndex, UDATA resolveFlags)
+{
+	J9RAMConstantDynamicRef *ramCPEntry = (J9RAMConstantDynamicRef*)ramCP + cpIndex;
+	J9JavaVM *vm = vmThread->javaVM;
+	j9object_t value = NULL;
+	j9object_t cpException = NULL;
+	bool resolved = false;
+
+retry:
+	value = ramCPEntry->value;
+	/* Fast path resolution for previously resolved entries */
+	if (NULL != value) {
+		resolved = true;
+	} else {
+		cpException = ramCPEntry->exception;
+		/* check if already resolved to NULL or exception */
+		if (NULL != cpException) {
+			J9Class *throwable = J9VMJAVALANGTHROWABLE_OR_NULL(vm);
+			J9Class *clazz = J9OBJECT_CLAZZ(vmThread, cpException);
+			if (cpException == vm->voidReflectClass->classObject) {
+				/* Resolved with null reference */
+				resolved = true;
+			} else if (isSameOrSuperClassOf(throwable, clazz)) {
+				/* Resolved with exception */
+				VM_VMHelpers::setExceptionPending(vmThread, cpException);
+				resolved = true;
+			}
+		}
+	}
+
+	if (!resolved) {
+		/* Slow path, attempt to acquire resolve mutex or retry, or wait */
+		omrthread_monitor_enter(vm->constantDynamicMutex);
+		value = ramCPEntry->value;
+		if (NULL != value) {
+			omrthread_monitor_exit(vm->constantDynamicMutex);
+			goto retry;
+		} else {
+			cpException = ramCPEntry->exception;
+			/* check if already resolved to NULL or exception */
+			if (NULL != cpException) {
+				J9Class *throwable = J9VMJAVALANGTHROWABLE_OR_NULL(vm);
+				J9Class *clazz = J9OBJECT_CLAZZ(vmThread, cpException);
+				if (cpException == vm->voidReflectClass->classObject) {
+					/* Resolved with null reference */
+					omrthread_monitor_exit(vm->constantDynamicMutex);
+					goto retry;
+				} else if (isSameOrSuperClassOf(throwable, clazz)) {
+					/* Resolved with exception */
+					omrthread_monitor_exit(vm->constantDynamicMutex);
+					goto retry;
+				} else if (cpException != vmThread->threadObject) {
+					/* Another thread is currently resolving this entry */
+					internalReleaseVMAccess(vmThread);
+					omrthread_monitor_wait(vm->constantDynamicMutex);
+					omrthread_monitor_exit(vm->constantDynamicMutex);
+					internalAcquireVMAccess(vmThread);
+					goto retry;
+				}
+			}
+		}
+
```

Comment:
add comment to indicate this thread is the one doing the resolve

Reply:
comment added